torch/csrc/distributed/rpc/rref.h

```diff
@@ -209,25 +207,32 @@ class RRef {
   // Returns true if this is the ``OwnerRRef``
   virtual bool isOwner() const = 0;
 
-  // returns true if this RRef holds an py::object, false if IValue
-  virtual bool isPyObj() = 0;
+  inline bool isPyObj() {
+    return type_ == PyObjectType::get();
+  }
+  inline const TypePtr type() {
+    return type_;
+  }
 
  protected:
   friend class RRefContext;
 
-  RRef(worker_id_t ownerId, const RRefId& rrefId);
+  RRef(worker_id_t ownerId, const RRefId& rrefId, const TypePtr& type);
 
   RRefForkData fork() const;
 
   const worker_id_t ownerId_;
   const RRefId rrefId_;
+
+  // type field to denote the type of the element that the RRef is holding
+  // it could be any TypePtr that JIT support, including PyObjectType
+  const TypePtr type_;
```

Comment:
Was wondering whether TypeKind would be sufficient here? Is there a reason we use TypePtr over TypeKind?